ReSharper.FSharp/src/FSharp.Psi.Services/src/Generate/GenerateProvider.fs

```diff
@@ -317,15 +324,24 @@ type FSharpOverridingMembersBuilder() =
         use disableFormatter = new DisableCodeFormatter()
 
         let typeDecl = context.Root :?> IFSharpTypeDeclaration
-        let typeRepr = typeDecl.TypeRepresentation
 
-        match typeRepr with
+        match typeDecl.TypeRepresentation with
         | :? IUnionRepresentation as unionRepr ->
             unionRepr.UnionCasesEnumerable
             |> Seq.tryHead
-            |> Option.iter EnumCaseLikeDeclarationUtil.addBarIfNeeded 
+            |> Option.iter EnumCaseLikeDeclarationUtil.addBarIfNeeded
+        | :? ITypeAbbreviationRepresentation as abbrRepr when abbrRepr.CanBeUnionCase ->
+            let factory = typeDecl.CreateElementFactory()
+            let caseName = FSharpNamingService.mangleNameIfNecessary abbrRepr.AbbreviatedTypeOrUnionCase.SourceName
+            let declGroup = factory.CreateModuleMember($"type U = | {caseName}")
+            let typeDeclaration = declGroup.FirstChild.As<IFSharpTypeDeclaration>()
```

Comment:
I'd use `:?>` since we generate the tree for the factory ourselves and we expect there to be this particular type.
Typed access like `declGroup.TypeDeclarations` is also something that should be used in the most cases, because it allows changing the tree grammar/structure much easier (and we do it from time to time).

```suggestion
            let typeDeclaration = declGroup.TypeDeclarations[0] :?> IFSharpTypeDeclaration
```